### ISSUE TITLE
Pattern matching on generic NamedTuple union loses specialized type parameter (T becomes object)

### ISSUE THREAD

**Issue by LoicRiegel** [astral-sh/ty#3916]
### Summary

Hi, I found this:

`main.py`:

```py
from typing import NamedTuple, reveal_type


class TaskResultSuccess[TData](NamedTuple):
    data: TData


class TaskResultError(NamedTuple):
    error_msg: str


class TaskResult[TData]:
    result: TaskResultSuccess[TData] | TaskResultError


def run(task_result: TaskResult[int]) -> None:
    match task_result.result:
        case TaskResultSuccess(data):
            reveal_type(data)  # expected: int, ty may reveal: object
            value: int = data  # may fail in ty if data is object
        case TaskResultError(error_msg):
            reveal_type(error_msg)  # expected: str

```

Diagnostics:
```
Revealed type: `object` (revealed-type) [Ln 19, Col 25]
Object of type `object` is not assignable to `int` (invalid-assignment) [Ln 20, Col 26]
Declared type [Ln 20, Col 20]
Revealed type: `str` (revealed-type) [Ln 22, Col 25]
```

`ty.json`:

```json
{
    "environment": {
        "python-version": "3.12"
    },
    "rules": {
        "undefined-reveal": "ignore"
    }
}
```

[Playground](https://play.ty.dev/268efe49-6282-4a10-a477-95c396cf129c)

### Version

ty 0.0.56 (3078f9fa2 2026-07-01)

### Other type checkers

Lokks like it's well handled by Pyright, but not mypy reveals Any.

**Comment by carljm** [comment]
This is #1578.

The result ty gives here is strictly correct, because multiple inheritance from both `TaskResultSuccess` and `TaskResultError` is possible. So `case TaskResultSuccess(data):` does not guarantee that the result is a `TaskResultSuccess[int]` -- it could be a `TaskResultError` which is also a `TaskResultSuccess[Whatever]`, which makes it part of the type `TaskResultSuccess[int] | TaskResultError` via the `TaskResultError` union element, while also still matching the `TaskResultSuccess` match arm.

The most straightforward way to express to ty that this is impossible, is to decorate one or both of `TaskResultSuccess` and `TaskResultError` as `@final`: https://play.ty.dev/8fbcd109-703e-4285-afc0-d43e021571b9

**Comment by LoicRiegel** [comment]
Hi, thank you very much for your response.

I hadn't considered the multi-inheritance case at all!
Reading the issue you linked was also super interesting.

In my case, I'm fine with using "@final", but I think the current situation still can be improved.

It would be great if ty could report that multiple inheritance is the reason for the loss of precision. Something like:
> Note: narrowing to TaskResultSuccess[T] is not possible because
TaskResultSuccess may participate in multiple inheritance. Consider marking involved classes as @final.

Also, what if I don't have control over the involved types, but I still want to build a sum type out of them?
I wonder whether some explicit opt-in mechanism (perhaps in ty_extensions) could let users declare that a particular union should be treated as disjoint for narrowing purposes, while keeping the default behavior sound.
Of course, such a mechanism would likely trade some soundness for ergonomics, so there may be downsides I'm not considering.

**Comment by carljm** [comment]
@LoicRiegel Thanks for the suggestions! We are definitely open to ideas for how to best balance correctness and usability here. We may just introduce a global strictness option that unsoundly considers many unions disjoint to better match the behavior of other type checkers.

**Comment by LoicRiegel** [comment]
I'll follow that closely then :) And thank you very much for the amazing work on ty!
